### PR TITLE
pkg/metrics: self-manage global expvar map

### PR DESCRIPTION
**Pull request by yichengq**
This helps the embedded tests.

### CODE REVIEW

**Comment by yichengq** [comment]
@xiang90 updated.

**Comment by yichengq** [comment]
This blocks #2140

File: pkg/metrics/metrics_test.go [hunk 1]
```diff
@@ -46,3 +73,14 @@ func TestMetrics(t *testing.T) {
 		t.Errorf("map after deletion = %s, want %s", m, w)
 	}
 }
+
+func reset() {
```
Comment (xiang90) [inline pkg/metrics/metrics_test.go]: how about moving this to metrics.go? 
we might make the func public.

File: pkg/metrics/metrics_test.go [hunk 2]
```diff
@@ -46,3 +73,14 @@ func TestMetrics(t *testing.T) {
...
+	vars = make(map[string]expvar.Var)
+	varKeys = nil
+}
+
+type stringVar struct {
```
Comment (xiang90) [inline pkg/metrics/metrics_test.go]: why do we need to have this stringVar?

File: pkg/metrics/metrics_test.go (hunk 1 above)
Comment (yichengq) [inline pkg/metrics/metrics_test.go]: I think we can do this when we want to export it.

File: pkg/metrics/metrics_test.go (hunk 2 above)
Comment (yichengq) [inline pkg/metrics/metrics_test.go]: removed

File: pkg/metrics/metrics_test.go (hunk 1 above)
Comment (xiang90) [inline pkg/metrics/metrics_test.go]: this should not belong to test anyway

Comment (yichengq) [inline pkg/metrics/metrics_test.go]: It helps to clean the environment that is polluted by tests

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -109,3 +90,75 @@ func (m *Map) String() string {
...
+// but existing entries may be concurrently updated.
+func Do(f func(expvar.KeyValue)) {
+	mutex.RLock()
+	defer mutex.RUnlock()
+	for _, k := range varKeys {
+		f(expvar.KeyValue{k, vars[k]})
+	}
+}
+
+func reset() {
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: // For test only

**Comment by xiang90** [comment]
lgtm

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (xiang90) [inline pkg/metrics/metrics.go]: you need to hold the lock to do the reset

**Pull request merged**